File: baseunits/modules/MangaTail.pas
unit MangaTail;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML;

implementation

uses FMDVars;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  s: String;
  v: IXQValue;
  i, x: Integer;
begin
  Result := NET_PROBLEM;
  s := Module.RootURL + '/directory';
  if AURL <> '0' then
    s += '?page=' + AURL;
  if MangaInfo.FHTTP.GET(s) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do try
      i := 1;
      for v in XPath('//ul[@class="pagination"]/li') do begin
        x := StrToIntDef(v.toString, 1);
        if x > i then i := x;
      end;
      updateList.CurrentDirectoryPageNumber := i - 1;
      XPathHREFAll('//table[contains(@class,"directory_list")]//tr/td[1]/a', ALinks, ANames);
      for i := 0 to ANames.Count - 1 do
      begin
        s := ANames[i];
        if RightStr(s, 6) = ' Manga' then
        begin
          SetLength(s, Length(s) - 6);
          ANames[i] := s;
        end;
      end;
    finally
      Free;
    end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  s: String;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do try
        coverLink := XPathString('//*[@class="bookface"]/img/@src');
        if title = '' then begin
          title := XPathString('//h1');
          if RightStr(title, 6) = ' Manga' then
            SetLength(title, Length(title) - 6);
        end;
        status := MangaInfoStatusIfPos(XPathString('//*[contains(@class,"field-status")]'));
        XPathHREFAll('//table[contains(@class,"chlist")]//tr/td[1]/a', chapterLinks, chapterName);
        InvertStrings([chapterLinks, chapterName]);
        s := XPathStringAll('//iframe[contains(.,"field-name")]/substring-after(substring-before(.,"""}}"),":en"":""")', LineEnding);
        if s <> '' then begin
          s := StringReplace(s, '\n', LineEnding, [rfReplaceAll]);
          s := StringReplace(s, '\', '', [rfReplaceAll]);
          ParseHTML(s);
          coverLink := XPathString('//img/@src');
          authors := XPathString('//*[contains(@class,"author")]/*[@class="field-items"]/string-join(.//text(),", ")');
          artists := XPathString('//*[contains(@class,"artist")]/*[@class="field-items"]/string-join(.//text(),", ")');
          genres := XPathString('//*[contains(@class,"genres")]/*[@class="field-items"]/string-join(.//text(),", ")');
          summary := XPathString('//*[contains(@class,"summary")]/*[@class="field-items"]//text()');
        end;
      finally
        Free;
      end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(MaybeFillHost(Module.RootURL, AURL + '?page=all')) then
    begin
      Result := True;
      XPathStringAll('//*[@id="images"]//img[not(contains(@src,"adsense"))]/@src', Document, PageLinks);
    end;
  end;
end;

procedure RegisterModule;

  function AddWebsiteModule(AWebsite, ARootURL: String): TModuleContainer;
  begin
    Result := AddModule;
    with Result do
    begin
      Website := AWebsite;
      RootURL := ARootURL;
      Category := 'English';
      OnGetNameAndLink := @GetNameAndLink;
      OnGetInfo := @GetInfo;
      OnGetPageNumber := @GetPageNumber;
    end;
  end;

begin
  AddWebsiteModule('MangaTail', 'http://www.mangatail.com');
  AddWebsiteModule('MangaSail', 'http://www.mangasail.com');
end;

initialization
  RegisterModule;

end.

File: lua/modules/MangaTail.lua
function getinfo()
  mangainfo.url=MaybeFillHost(module.RootURL, url)
  http.cookies.values['has_js'] = '1'
  if http.get(mangainfo.url) then
    local x=TXQuery.Create(http.document)
    if mangainfo.title == '' then
      mangainfo.title = string.gsub(x.xpathstring('//h1[@class="page-header"]'), ' Manga$', '')
    end
    mangainfo.status = MangaInfoStatusIfPos(x.xpathstring('//*[contains(@class,"field-status")]'))
    x.xpathhrefall('//table[contains(@class,"chlist")]//tr/td[1]/a', mangainfo.chapterlinks, mangainfo.chapternames)
    InvertStrings(mangainfo.chapterlinks,mangainfo.chapternames)    
    
    local s = x.xpathstring('//script[contains(., "jQuery.extend(Drupal")]')
    s = GetBetween('settings,', ');', s)
    x.parsehtml(s)    
    local v = x.xpath('json(*)/authcacheP13nAjaxAssemblies')
    local summaryQuery = x.xpathstring('./span.authcache-p13n-asm-field-node-body', v)
    local artistQuery = x.xpathstring('./span.authcache-p13n-asm-field-node-field-artist', v)
    local authorQuery = x.xpathstring('./span.authcache-p13n-asm-field-node-field-author', v)
    local genresQuery = x.xpathstring('./span.authcache-p13n-asm-field-node-field-genres', v)
    local coverQuery = x.xpathstring('./span.authcache-p13n-asm-field-node-field-image2', v)
    local statusQuery = x.xpathstring('./span.authcache-p13n-asm-field-node-field-status', v)
    
    function getField(aurl, query)
      if aurl ~= '' then
        http.reset()
        http.headers.values['X-Authcache'] = '1'
        if http.xhr(module.rooturl .. aurl) then
          local s = GetBetween(':"', '"}', StreamToString(http.document))
          x.parsehtml(s:gsub('\\"', '"'):gsub('\\/', '/'))
          return x.xpathstring(query)
        end
      end
      return ''
    end
    
    mangainfo.summary = getField(summaryQuery, '*')
    mangainfo.authors = getField(authorQuery, '//div[contains(@class, "field-item")]')
    mangainfo.artists = getField(artistQuery, '//div[contains(@class, "field-item")]')
    mangainfo.coverlink = getField(coverQuery, '//img/@src')
    mangainfo.genres = getField(genresQuery, 'string-join(//a, ", ")')

    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  task.pagelinks.clear()
  if http.get(MaybeFillHost(module.rooturl, url .. '?page=all')) then
    local x=TXQuery.Create(http.Document)
    x.xpathstringall('//*[@id="images"]//img[not(contains(@src,"adsense"))]/@src', task.pagelinks)
  else
    return false
  end
  return true
end

function getnameandlink()
  local s = module.rooturl .. '/directory'
  if url ~= '0' then s = s .. '?page=' .. url; end
  if http.get(s) then
    local x = TXQuery.Create(http.Document)
    local i = 1
    local v = x.xpath('//ul[@class="pagination"]/li')
    for j = 1, v.count do
      local v1 = v.get(j)
      local x = tonumber(v1.toString)
      if (x ~= nil) and (x > i) then i = x; end
    end
    updatelist.CurrentDirectoryPageNumber = i - 1
    v = x.xpath('//table[contains(@class,"directory_list")]//tr/td[1]/a')
    for j = 1, v.count do
      local v1 = v.get(j)
      local s = v1.toString
      if string.match(s:upper(), ' MANGA$') ~= nil then
        s = s:sub(1, -7)
      end
      links.add(v1.getAttribute('href'))
      names.add(s)
    end
    return no_error
  else
    return net_problem
  end
end

function AddWebsiteModule(site, url)
  local m=NewModule()
  m.category='English'
  m.website=site
  m.rooturl=url
  m.lastupdated='April 5, 2018'
  m.ongetinfo='getinfo'
  m.ongetpagenumber='getpagenumber'
  m.ongetnameandlink='getnameandlink'
  return m
end

function Init()
  AddWebsiteModule('MangaTail', 'https://www.mangatail.me')
  AddWebsiteModule('MangaSail', 'https://www.mangasail.co')
end 